amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java

```diff
@@ -296,45 +309,27 @@ private void addArrivedRecordsInput(PrefetchRecordsRetrieved recordsRetrieved) t
                 throw new PositionResetException();
             }
         }
-        prefetchCounters.added(recordsRetrieved.processRecordsInput);
-    }
-
-    /**
-     * Method that will be called by the 'publisher thread' and the 'demand notifying thread',
-     * to drain the events if the 'event notifying thread' do not have the control.
-     */
-    private synchronized void drainQueueForRequestsIfAllowed() {
-        if (!shouldDrainEventOnlyOnAck.get()) {
-            drainQueueForRequests();
-        }
+        publisherSession.prefetchCounters().added(recordsRetrieved.processRecordsInput);
     }
 
     /**
      * Method to drain the queue based on the demand and the events availability in the queue.
      */
     private synchronized void drainQueueForRequests() {
-        final RecordsRetrieved recordsToDeliver = peekNextResult();
+        final PrefetchRecordsRetrieved recordsToDeliver = (PrefetchRecordsRetrieved) peekNextResult();
         // If there is an event available to drain and if there is at least one demand,
         // then schedule it for delivery
-        if (requestedResponses.get() > 0 && recordsToDeliver != null) {
+        if (publisherSession.hasDemandToPublish() && canDispatchRecord(recordsToDeliver)) {
+            subscriber.onNext(recordsToDeliver.prepareForPublish());
+            recordsToDeliver.dispatched();
             lastEventDeliveryTime = Instant.now();
-            subscriber.onNext(recordsToDeliver);
-            if (!shouldDrainEventOnlyOnAck.get()) {
-                log.debug("{} : Notifier thread takes over the draining control. Queue Size : {}, Demand : {}", shardId,
-                        getRecordsResultQueue.size(), requestedResponses.get());
-                shouldDrainEventOnlyOnAck.set(true);
-            }
-        } else {
-            // Since we haven't scheduled the event delivery, give the drain control back to publisher/demand-notifier
-            // thread.
-            if (shouldDrainEventOnlyOnAck.get()) {
-                log.debug("{} : Publisher thread takes over the draining control. Queue Size : {}, Demand : {}",
-                        shardId, getRecordsResultQueue.size(), requestedResponses.get());
-                shouldDrainEventOnlyOnAck.set(false);
-            }
         }
     }
 
+    private static boolean canDispatchRecord(PrefetchRecordsRetrieved recordsToDeliver) {
```

Comment:
worth adding a comment here, or in the PrefetchesRecordsRetrieved class about multi-threading, and this flag protecting against delivering records twice.

Reply:
done